Refactor Statistics to Doctrine [MAILPOET-2503]

amine-mp opened the pull request:
[MAILPOET-2503]

[MAILPOET-2503]: https://mailpoet.atlassian.net/browse/MAILPOET-2503

costasovo commented on lib/API/JSON/ResponseBuilders/NewslettersResponseBuilder.php:
```diff
@@ -27,12 +44,36 @@ public function build(NewsletterEntity $newsletter) {
+        $data['totalSent'] = $this->newslettersStatsRepository->getTotalSentCount($newsletter);
+      }
+      if ($relation === self::RELATION_CHILDREN_COUNT) {
+        $data['childrenCount'] = count($newsletter->getChildren());
```
This will require to fetch and hydrate all children entities. What about creating a getChildrenCount() method which would only query DB for the count on newslettersRepository?

costasovo commented on lib/Entities/StatisticsWooCommercePurchaseEntity.php:
```diff
@@ -0,0 +1,64 @@
+   * @ORM\JoinColumn(name="subscriber_id", referencedColumnName="id")
+   * @var SubscriberEntity|null
+   */
+  private $subscriber;
```
I noticed that when we delete a subscriber we don't delete his purchases. I'm not sure what happens to this entity when the association refers to non-existing entity.

costasovo replied:
It is probably ok for now since we don't work with the entity directly. It seems that when we delete subscriber (possible even other entities) we don't clean all related records and we leave some foreign keys. This may not pay off in the future. It makes sense to keep some anonymous statistic data after we delete a subscriber. Maybe we should make those relations nullable in DB and cleanup foreign keys when we delete a subscriber. So this is probably a more general problem not related directly to this PR.

amine-mp commented on lib/Newsletter/Statistics/NewsletterStatistics.php:
```diff
@@ -16,11 +16,15 @@ class NewsletterStatistics {
   private $totalSentCount;
 
-  public function __construct($clickCount, $openCount, $unsubscribeCount, $totalSentCount) {
+  /** @var NewsletterWooCommerceRevenue|null */
```
It's used here https://github.com/mailpoet/mailpoet/pull/2558/files/c7ff32e1966256aa17a958ff3c2a23caf7e8cb49..72e4f11db8106de596694489b78ec2b72a1d68d7#diff-7915794b23c9896cc65d74576b61cfcaR62

amine-mp commented on lib/Entities/StatisticsWooCommercePurchaseEntity.php:
```diff
@@ -0,0 +1,64 @@
+  /**
+   * @ORM\ManyToOne(targetEntity="MailPoet\Entities\NewsletterEntity")
+   * @ORM\JoinColumn(name="newsletter_id", referencedColumnName="id")
+   * @var NewsletterEntity|null
```
Ok, I added the constructor.

amine-mp commented:
The goal of this task was to get rid of Paris models: `StatisticsOpens`, `StatisticsClicks` and `StatisticsUnsubscribe`, and replace them with Doctrine entities. After some refactoring, I noticed that we need to refactor `API\JSON\Newsletters::listing` to use Doctrine and that would take more time so we decided to do it in a separated task https://mailpoet.atlassian.net/browse/MAILPOET-2655

amine-mp commented:
I fixed the bug in the stats (is was a camelCase vs underscores issue) and I added a test.
For premium plugin, it will require rewriting `CampaignStats::getClickedLinks`, do you think we need to do that as part of this PR?

costasovo commented:
@amine-mp I think it would be better to do the premium plugin part in a separate task so that we don't block this one. Could you please create the ticket?

amine-mp commented:
I created the task https://mailpoet.atlassian.net/browse/PREMIUM-142
